### ISSUE TITLE
Fasta crash in strong mode when when mixins are partially from dill file

### ISSUE THREAD

**Issue by jensjoha** [dart-lang/sdk#32318]
Originally reported as https://github.com/flutter/flutter/issues/14850.

Reproduction created in https://dart-review.googlesource.com/43723.

What's happening is that file1.dart is recompiled from source and contains a class like "A extends Object with B<Object>, C" where B and C are in another file (say file2.dart) that was loaded from a dill file.
This can for instance happens when bootstrapping the incremental compiler from a dill file, which is the case for flutter.

**Comment by kmillikin** [comment]
The crash is because `isSyntheticMixinImplementation` doesn't mean exactly what I thought it meant.  (Or maybe it does and it's not implemented correctly.)  I'm not yet sure why the crash doesn't happen on the initial compilation but I'll investigate.

I can fix the crash but there is a deeper issue that also needs to be fixed: mixin canonicalization is a local transformation, but mixin type argument inference relies on the global class hierarchy.  What's happening is that we start with code like this:

```dart
// In library lib1
abstract class C<T> extends Object with B<T> {}
abstract class B<T> extends Object {}

// In library lib2
class A extends Object with B<Object>, C {}
```

We canonicalize the mixin application Object with B by introducing library-private classes like this (partial result of the transformation):

```dart
// In library lib1
abstract class _Object&B<T> = Object with B<T> {}
abstract class C<T> extends _Object&B<T> {}
abstract class B<T> extends Object {}

// In library lib2
abstract class _Object&B<T> = Object with B<T> {}
class A extends _Object&B<Object> with C {}
```

To infer the type arguments for C in the mixin application in class A, we try to match the superclass lib2._Object&B<Object> against mixin C's superclass constraint which is lib1._Object&B<T> and these are two different classes.

**Comment by kmillikin** [comment]
The reason that this only occurs for classes that were loaded from VM-specific .dill files is that mixin application classes have been translated into normal classes.  In that case we do not have a mixed-in type but we know that the only implemented interface is the original mixed-in type.

The issue with library privacy was a symptom, not the cause.  Proposed fix is [here](https://dart-review.googlesource.com/c/sdk/+/43965).

**Comment by kmillikin** [comment]
Fixed in 13067387a4d549aafa971dc70a2e408686bac21d.
